test/model/src/include/model/kv_table.h

```diff
@@ -0,0 +1,170 @@
+/*-
+ * Public Domain 2014-present MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef MODEL_KV_TABLE_H
+#define MODEL_KV_TABLE_H
+
+#include <atomic>
+#include <map>
+#include <mutex>
+#include <string>
+
+#include "model/data_value.h"
+#include "model/kv_table_item.h"
+#include "model/kv_update.h"
+#include "model/verify.h"
+#include "wiredtiger.h"
+
+namespace model {
+
+/*
+ * kv_table --
+ *     A database table with key-value pairs.
+ */
+class kv_table {
+
+public:
+    /*
+     * kv_table::kv_table --
+     *     Create a new instance.
+     */
+    inline kv_table(const char *name) : _name(name) {}
+
+    /*
+     * kv_table::name --
+     *     Get the name of the table.
+     */
+    inline const char *
+    name() const noexcept
```

Comment:
This expose internal data to a different timeline. Have you considered either `string_view` or a smart pointer?

Reply:
This is also somewhat of a judgment call. In the past, I've been usually fine providing read-only access to read-only internal state through pointers. The only concern that I would have is what if someone holds on to the pointer after the table object goes away—I wouldn't normally worry about that, as it is a long-lived object.

I guess `c_str()` was a poor version of `std::string_view`, as neither actually creates a copy of the string, but both point to the internal data of the string. Between the two, I'm somewhat more in favor of keeping it as `c_str()`, so that the returned value is usable in any WiredTiger calls—or any calls from C as this may be eventually called from a C test program—that require `const char*`. It is unfortunate that `std::string_view` does not provide a `c_str()` equivalent, but it makes sense as it does not store a NULL-terminated string internally.

Smart pointer to `std::string` might make sense if we are concerned that the caller may hold on to the pointer past the object's lifetime without making a copy. I suppose there is some value in that, but I'm not sure it's necessary.

I actually ran into the following problem the other day—I was doing something like this when setting up a WT cursor (I wasn't actually getting the name, but I'm just putting it here to illustrate the example):
```
cursor->set_key(cursor, t.name().c_str());
```
When I had `t.name()` return `std::string` (as opposed to a reference), I actually got undefined behavior. The reason was that the returned string fell out of scope and was destructed before WT used the provided pointer, which then pointed to unallocated memory. Returning a shared pointer would not necessarily solve the problem either, because the shared pointer object would also fall out of scope. I mean, of course the right solution would be to make a local copy of the string (or the shared pointer, if I ended up switching to that), but that's easy to forget—and then hard to figure out why the program was not behaving. (I wasted quite a bit of time debugging this.)

So actually tying the lifetime of the `const char*` pointer to the lifetime of `t` to me provides the most intuitive  behavior with the least chance of bugs—whether by returning `const char*` directly or by returning a read-only reference to the internal state.

If this was interacting with API written in C++ that understands C++ strings and/or shared pointers, it would be a different story. But WiredTiger's APIs are in C, so I still feel like referencing internal state is probably the best way to do this. And we're returning a const pointers to a read-only property, so it should be safe, of course providing that the lifetimes make sense as explained above.

But that's a very good question.

Reply:
I can appreciate that supporting interfacing with the WT C API adds an additional burden, and an accessor method like this is most useful.
Does it make sense to keep this and add to the comment that it is specifically for interfacing with the C API and should only be used in known limit scope situations, ie:
```
some_c_api(kv_table_instance.name());
```
And add an additional method that returns a value that can be capture for unspecified lifetime be it a copy, or some smart thingy?

Reply:
I added a comment explaining the lifetime of the returned pointer. I'll see if I need a version with a more C++ friendly return value in the future, but I think this should be at least good enough for now!

Reply:
Agreed